File: apps/api/app/shared/domain/money.py
"""Money value object.

Amounts are stored as an integer number of cents to avoid floating-point
rounding errors, which would silently corrupt financial totals.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class Money:
    amount_cents: int
    currency: str

    def __post_init__(self) -> None:
        if isinstance(self.amount_cents, bool) or not isinstance(self.amount_cents, int):
            raise TypeError("amount_cents must be an int")
        if self.amount_cents < 0:
            raise ValueError("amount_cents cannot be negative")
        if len(self.currency) != 3 or not self.currency.isalpha():
            raise ValueError("currency must be a 3-letter ISO 4217 code")
        object.__setattr__(self, "currency", self.currency.upper())

File: apps/api/internal/shared/domain/money.go
package domain

import (
	"errors"
	"strings"
	"unicode"
)

type Money struct {
	AmountCents int64
	Currency    string
}

func NewMoney(amountCents int64, currency string) (Money, error) {
	currency = strings.ToUpper(strings.TrimSpace(currency))
	if amountCents < 0 {
		return Money{}, errors.New("amount_cents cannot be negative")
	}
	if len(currency) != 3 {
		return Money{}, errors.New("currency must be a 3-letter ISO 4217 code")
	}
	for _, char := range currency {
		if !unicode.IsLetter(char) {
			return Money{}, errors.New("currency must be a 3-letter ISO 4217 code")
		}
	}
	return Money{AmountCents: amountCents, Currency: currency}, nil
}
